File: java/src/main/java/com/newnancity/newnanmanager/services/TownService.java
package com.newnancity.newnanmanager.services;

import com.newnancity.newnanmanager.exceptions.NewNanManagerException;
import com.newnancity.newnanmanager.generated.api.TownServiceApi;
import com.newnancity.newnanmanager.generated.model.CreateTownRequest;
import com.newnancity.newnanmanager.generated.model.ListTownsResponse;
import com.newnancity.newnanmanager.generated.model.Town;
import com.newnancity.newnanmanager.generated.model.TownDetailResponse;
import com.newnancity.newnanmanager.generated.model.UpdateTownRequest;

/** Town management service. */
public final class TownService extends AbstractService {
    private final TownServiceApi api;

    public TownService(TownServiceApi api) {
        this.api = api;
    }

    public Town createTown(CreateTownRequest request) throws NewNanManagerException {
        return execute(() -> api.createTown(request).execute());
    }

    public TownDetailResponse getTown(int id) throws NewNanManagerException {
        return getTown(id, null);
    }

    public TownDetailResponse getTown(int id, Boolean detail) throws NewNanManagerException {
        return execute(() -> api.getTown(id).detail(detail).execute());
    }

    public Town updateTown(int id, UpdateTownRequest request) throws NewNanManagerException {
        return execute(() -> api.updateTown(id, request).execute());
    }

    public void deleteTown(int id) throws NewNanManagerException {
        executeVoid(() -> api.deleteTown(id).execute());
    }

    public ListTownsResponse listTowns() throws NewNanManagerException {
        return listTowns(null, null, null, null, null, null);
    }

    public ListTownsResponse listTowns(
            Integer page,
            Integer pageSize,
            String name,
            String search,
            Integer minLevel,
            Integer maxLevel) throws NewNanManagerException {
        return execute(() -> api.listTowns()
                .page(page)
                .pageSize(pageSize)
                .name(name)
                .search(search)
                .minLevel(minLevel)
                .maxLevel(maxLevel)
                .execute());
    }

    public TownServiceApi raw() {
        return api;
    }
}

File: kotlin/src/main/kotlin/com/nanmanager/bukkit/services/TownService.kt
package com.nanmanager.bukkit.services

import com.nanmanager.bukkit.http.HttpClient
import com.nanmanager.bukkit.models.*

/** 城镇管理服务；传输和模型解析由 OpenAPI 生成客户端负责。 */
class TownService(httpClient: HttpClient) : BaseService(httpClient) {
    private val api get() = httpClient.generatedApis.towns

    fun createTown(request: CreateTownRequest): Town {
        val body = toGenerated<com.newnanmanager.generated.models.CreateTownRequest>(request)
        return generated<Town, com.newnanmanager.generated.models.Town> {
            api.createTownWithHttpInfo(body)
        }
    }

    fun getTown(request: GetTownRequest): TownDetailResponse {
        return generated<TownDetailResponse, com.newnanmanager.generated.models.TownDetailResponse> {
            api.getTownWithHttpInfo(request.id, request.detail)
        }
    }

    fun updateTown(request: UpdateTownRequest): Town {
        val body = toGenerated<com.newnanmanager.generated.models.UpdateTownRequest>(request)
        return generated<Town, com.newnanmanager.generated.models.Town> {
            api.updateTownWithHttpInfo(request.id, body)
        }
    }

    fun deleteTown(request: DeleteTownRequest) {
        generatedVoid { api.deleteTownWithHttpInfo(request.id) }
    }

    fun listTowns(request: ListTownsRequest): ListTownsResponse {
        return generated<ListTownsResponse, com.newnanmanager.generated.models.ListTownsResponse> {
            api.listTownsWithHttpInfo(
                page = request.page,
                pageSize = request.pageSize,
                name = request.name,
                search = request.search,
                minLevel = request.minLevel,
                maxLevel = request.maxLevel
            )
        }
    }
}
